discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -94,55 +106,125 @@ def refresh_user_balance(session: Session, redis: Redis, user_bank_acct: str):
+        instructions = tx_message["instructions"]
+        user_bank_program_index = account_keys.index(USER_BANK_ADDRESS)
+        for instruction in instructions:
+            if instruction["programIdIndex"] == user_bank_program_index:
+                return instruction
+
+        return None
+    except Exception as e:
+        logger.error(
+            f"index_user_bank.py | Error processing instruction valid, {e}",
+            exc_info=True,
+        )
+        return None
 
 def process_user_bank_tx_details(
-    session: Session, redis: Redis, tx_info, tx_sig, timestamp
+    session: Session,
+    redis: Redis,
+    tx_info,
+    tx_sig,
+    timestamp
 ):
-    meta = tx_info["result"]["meta"]
+    result: TransactionInfoResult = tx_info["result"]
+    meta = result["meta"]
     error = meta["err"]
     if error:
         logger.info(
             f"index_user_bank.py | Skipping error transaction from chain {tx_info}"
         )
         return
     account_keys = tx_info["result"]["transaction"]["message"]["accountKeys"]
-    instructions = meta["logMessages"]
-    for msg in instructions:
-        if "EthereumAddress" in msg:
-            public_key_str, public_key_bytes = parse_eth_address_from_msg(msg)
-            logger.info(f"index_user_bank.py | {public_key_str}")
-            # Rederive address based on user public key
-            _, derived_address = get_address_pair(
-                WAUDIO_PROGRAM_PUBKEY, public_key_bytes, USER_BANK_KEY, SPL_TOKEN_ID_PK
-            )
-            bank_acct = str(derived_address[0])
+    tx_message = result["transaction"]["message"]
+
+    # Check for valid instruction
+    has_create_token_instruction = any(
+        log == "Program log: Instruction: CreateTokenAccount"
+        for log in meta["logMessages"]
+    )
+    has_claim_instruction = any(
+        log == "Program log: Instruction: Claim"
+        for log in meta["logMessages"]
+    )
+
+    if not has_create_token_instruction and not has_claim_instruction:
+        return
+
+    instruction = get_valid_instruction(tx_message, meta)
+    if instruction is None:
+        logger.error(f"index_user_bank.py | {tx_sig} No Valid instruction found")
+        return
+
+    if has_create_token_instruction:
+        tx_data = instruction["data"]
+        parsed_token_data = parse_create_token_data(tx_data)
+        eth_addr = parsed_token_data["eth_address"]
+        decoded = base58.b58decode(tx_data)[1:]
+        public_key_bytes = decoded[:20]
+        _, derived_address = get_address_pair(
+            WAUDIO_PROGRAM_PUBKEY,
+            public_key_bytes,
+            USER_BANK_KEY,
+            SPL_TOKEN_ID_PK
+        )
+        bank_acct = str(derived_address[0])
+        try:
             # Confirm expected address is present in transaction
-            try:
-                bank_acct_index = account_keys.index(bank_acct)
+            bank_acct_index = account_keys.index(bank_acct)
```

Comment:
if the index is zero, will the check below fail?

Reply:
no, i think it returns a ValueError instead

Reply:
it's actually explicitly handled here

Reply:
https://github.com/AudiusProject/audius-protocol/pull/1849/files#diff-58f43296a3513f8dacca1b49a26e650ec743e1624002495e73d9ecb853060d5fR216